File: backend/python/primeshape/shape_templates.py
import cv2
import numpy as np


def reference_contours():
    shapes = {}
    for tips in (5, 6):
        for inner in (0.38, 0.5):
            t = np.arange(2 * tips) * np.pi / tips - np.pi / 2
            r = np.where(np.arange(2 * tips) % 2, inner, 1)
            shapes[f"Estrela de {tips} pontas:{inner}"] = np.c_[r * np.cos(t), r * np.sin(t)]
    for arm in (0.25, 0.38):
        shapes[f"Cruz:{arm}"] = np.array([[-arm,-1],[arm,-1],[arm,-arm],[1,-arm],[1,arm],[arm,arm],[arm,1],[-arm,1],[-arm,arm],[-1,arm],[-1,-arm],[-arm,-arm]])
    t = np.linspace(0, 2 * np.pi, 160, endpoint=False)
    shapes["Coração"] = np.c_[16 * np.sin(t) ** 3, -(13 * np.cos(t) - 5 * np.cos(2*t) - 2 * np.cos(3*t) - np.cos(4*t))] / 17
    t = np.linspace(0, np.pi, 100)
    shapes["Semicírculo"] = np.c_[np.cos(t), -np.sin(t)]
    for stem in (0.25, 0.4):
        shapes[f"Seta:{stem}"] = np.array([[-1,-stem],[0,-stem],[0,-0.85],[1,0],[0,0.85],[0,stem],[-1,stem]])
    shapes["Chevron"] = np.array([[-1,-1],[-0.1,-1],[1,0],[-0.1,1],[-1,1],[0,0]])
    return {name: np.round((points + 1.5) * 100).astype(np.int32).reshape(-1, 1, 2) for name, points in shapes.items()}


def mask_for(contour, size=48):
    points = contour[:, 0, :].astype(float)
    low, high = points.min(axis=0), points.max(axis=0)
    span = max(float((high - low).max()), 1)
    points = (points - (low + high) / 2) * ((size - 6) / span) + size / 2
    mask = np.zeros((size, size), dtype=np.uint8)
    cv2.fillPoly(mask, [np.round(points).astype(np.int32)], 1)
    return mask.astype(bool)


class TemplateMatcher:
    def __init__(self):
        self.references = reference_contours()
        self.bank = []
        self.labels = []
        for name, contour in self.references.items():
            center = contour[:, 0, :].mean(axis=0)
            for degrees in range(0, 360, 10):
                radians = np.deg2rad(degrees)
                rotation = np.array([[np.cos(radians), -np.sin(radians)], [np.sin(radians), np.cos(radians)]])
                rotated = (contour[:, 0, :] - center) @ rotation.T
                self.bank.append(mask_for(rotated.reshape(-1, 1, 2)))
                self.labels.append(name.split(":")[0])
        self.bank = np.asarray(self.bank)

    def classify(self, contour):
        target = mask_for(contour)
        intersections = np.logical_and(self.bank, target).sum(axis=(1, 2))
        unions = np.logical_or(self.bank, target).sum(axis=(1, 2))
        scores = intersections / np.maximum(unions, 1)
        ranked = {}
        for label, score in zip(self.labels, scores):
            ranked[label] = max(ranked.get(label, 0), float(score))
        ordered = sorted(ranked.items(), key=lambda x: x[1], reverse=True)
        label, best = ordered[0]
        if best < 0.84 or best - ordered[1][1] < 0.055:
            return None
        return label, best, 0

File: frontend/js/browser/shape-templates.js
import { bounds } from "./geometry.js";

export function references() {
  const shapes = [];
  for (const tips of [5,6]) for (const inner of [0.38,0.5]) {
    shapes.push([`Estrela de ${tips} pontas`, Array.from({ length: 2 * tips }, (_, i) => {
      const t = i * Math.PI / tips - Math.PI / 2, r = i % 2 ? inner : 1;
      return [r * Math.cos(t), r * Math.sin(t)];
    })]);
  }
  for (const a of [0.25,0.38]) shapes.push(["Cruz", [[-a,-1],[a,-1],[a,-a],[1,-a],[1,a],[a,a],[a,1],[-a,1],[-a,a],[-1,a],[-1,-a],[-a,-a]]]);
  shapes.push(["Coração", Array.from({ length: 160 }, (_, i) => {
    const t = i * Math.PI / 80;
    return [16 * Math.sin(t) ** 3 / 17, -(13 * Math.cos(t) - 5 * Math.cos(2*t) - 2 * Math.cos(3*t) - Math.cos(4*t)) / 17];
  })]);
  shapes.push(["Semicírculo", Array.from({ length: 100 }, (_, i) => [Math.cos(i * Math.PI / 99), -Math.sin(i * Math.PI / 99)])]);
  for (const s of [0.25,0.4]) shapes.push(["Seta", [[-1,-s],[0,-s],[0,-0.85],[1,0],[0,0.85],[0,s],[-1,s]]]);
  shapes.push(["Chevron", [[-1,-1],[-0.1,-1],[1,0],[-0.1,1],[-1,1],[0,0]]]);
  return shapes;
}

function mask(points) {
  const [x,y,w,h] = bounds(points), scale = 28 / Math.max(w,h,1e-6);
  const normalized = points.map(p => [(p[0] - x - w/2) * scale + 16, (p[1] - y - h/2) * scale + 16]);
  const rows = new Uint32Array(32);
  for (let row = 0; row < 32; row++) {
    const scan = row + 0.5, intersections = [];
    for (let i = 0; i < normalized.length; i++) {
      const a = normalized[i], b = normalized[(i + 1) % normalized.length];
      if ((a[1] > scan) !== (b[1] > scan)) intersections.push(a[0] + (scan - a[1]) * (b[0] - a[0]) / (b[1] - a[1]));
    }
    intersections.sort((a,b) => a-b);
    for (let i = 0; i + 1 < intersections.length; i += 2) {
      for (let col = Math.max(0, Math.ceil(intersections[i] - 0.5)); col < Math.min(32, Math.ceil(intersections[i+1] - 0.5)); col++) rows[row] |= 1 << col;
    }
  }
  return rows;
}

function popcount(value) {
  value -= (value >>> 1) & 0x55555555;
  value = (value & 0x33333333) + ((value >>> 2) & 0x33333333);
  return (((value + (value >>> 4)) & 0x0f0f0f0f) * 0x01010101) >>> 24;
}

export class TemplateMatcher {
  constructor() {
    this.bank = [];
    for (const [label, points] of references()) for (let angle = 0; angle < 360; angle += 10) {
      const t = angle * Math.PI / 180;
      this.bank.push([label, mask(points.map(([x,y]) => [x * Math.cos(t) - y * Math.sin(t), x * Math.sin(t) + y * Math.cos(t)]))]);
    }
  }
  classify(points) {
    const target = mask(points), scores = new Map();
    for (const [label, reference] of this.bank) {
      let intersection = 0, union = 0;
      for (let i = 0; i < 32; i++) { intersection += popcount(target[i] & reference[i]); union += popcount(target[i] | reference[i]); }
      scores.set(label, Math.max(scores.get(label) || 0, intersection / Math.max(union, 1)));
    }
    const ordered = [...scores].sort((a,b) => b[1]-a[1]);
    return ordered[0][1] >= 0.84 && ordered[0][1] - ordered[1][1] >= 0.055 ? [...ordered[0], 0] : null;
  }
}
